scrapy/exporters.py

```diff
@@ -253,12 +259,23 @@ def _build_row(self, values):
     def _write_headers_and_set_fields_to_export(self, item):
         if self.include_headers_line:
             if not self.fields_to_export:
+                self._is_data_loss_warning_displayed = True
                 if isinstance(item, dict):
                     # for dicts try using fields of the first item
                     self.fields_to_export = list(item.keys())
                 else:
                     # use fields declared in Item
                     self.fields_to_export = list(item.fields.keys())
+            elif not self._fail_on_dataloss_warned:
+                if isinstance(item, dict):
+                    if len(item.keys()) > self.fields_to_export:
```

Comment:
Related to https://github.com/scrapy/scrapy/issues/4002#issuecomment-536662554:
I think these checks should compare the actual fields, not the amount of fields: `len({"a", "b"}) == len({"b", "c"})` but `{"a", "b"} != {"b", "c"}`.

Reply:
As per my understanding, if the `fields_to_export` is set and if some of the rows have more number of columns than the columns in `fields_to_export` then we should be showing the warning. That is the reason why I check for the length, if on any row the length of the row is greater than the length of the `fields_to_export`, I was showing the warning.

You are right, I will add the `len(fields_to_export)` to compare instead of comparing directly with `fields_to_export` which is not a number.

Reply:
The number is not enough, because as with the example that @elacuesta provided (`{"a", "b"} != {"b", "c"}`), two items may have the same number of columns but different field names.